baybe/utils/augmentation.py

```diff
@@ -0,0 +1,242 @@
+        +---+---+---+---+
+        | 0 | 3 | 5 | y |
+        +---+---+---+---+
+        | 0 | 4 | 5 | y |
+        +---+---+---+---+
+
+    *   Result with ``causing = ("A", [0, 1])`, `affected = [("B", [2,3])]``
+
+        +---+---+---+---+
+        | A | B | C | D |
+        +===+===+===+===+
+        | 0 | 2 | 5 | y |
+        +---+---+---+---+
+        | 1 | 3 | 5 | z |
+        +---+---+---+---+
+        | 0 | 3 | 5 | y |
+        +---+---+---+---+
+        | 1 | 2 | 5 | z |
+        +---+---+---+---+
+
+    *   Result with ``causing = ("A", [0])`, `affected = [("B", [2,3]), ("C", [5, 6])]``
+
+        +---+---+---+---+
+        | A | B | C | D |
+        +===+===+===+===+
+        | 0 | 2 | 5 | y |
+        +---+---+---+---+
+        | 1 | 3 | 5 | z |
+        +---+---+---+---+
+        | 0 | 3 | 5 | y |
+        +---+---+---+---+
+        | 0 | 2 | 6 | y |
+        +---+---+---+---+
+        | 0 | 3 | 6 | y |
+        +---+---+---+---+
+
+    Args:
+        df: The dataframe that should be augmented.
+        causing: Causing column name and its causing values.
+        affected: Affected columns and their invariant values.
+
+    Returns:
+        The augmented dataframe containing the original one.
+    """
+    new_rows: list[pd.DataFrame] = []
+
+    # Iterate through all rows that have a causing value in the respective column.
+    col_causing, vals_causing = causing
+    df_filtered = df.loc[df[col_causing].isin(vals_causing), :]
+    for _, row in df_filtered.iterrows():
+        # Augment the  specific row by growing a dataframe iteratively going through
+        # the affected columns. In each iteration augmented rows with that column
+        # changed to all possible values are added. If there is more than one affected
+        # column, it is important to include the augmented rows stemming from the
+        # preceding columns as well.
+        original_row = pd.DataFrame([row])
+
+        currently_added = original_row.copy()  # Start with the original row
+        for col_affected, vals_invariant in affected:
+            to_add = []
+
+            # Go through all previously added rows + the original row
+            for _, temp_row in currently_added.iterrows():
+                to_add += [
+                    new_row
+                    for val in vals_invariant
+                    if not _row_in_df(
+                        new_row := temp_row.pipe(lambda x: pd.DataFrame([x])).assign(
+                            **{col_affected: val}
+                        ),  # this takes the current row and replaces the affected value
+                        currently_added,
+                    )
+                ]
+            # Update the currently added rows
+            currently_added = pd.concat([currently_added] + to_add)
+
+        # Drop first entry because it's the original row and store added rows
+        currently_added = currently_added.iloc[1:, :]
+        new_rows.append(currently_added)
```

Comment:
I might be missing something, but I think this could be done easier? Just take the product of all affected columns' values and augment that? Something like this:

```suggestion
    affected_cols, affected_inv_vals = zip(*affected)
    affected_inv_vals_combinations = product(*affected_inv_vals)
    
    for _, r in df_filtered.iterrows():
        to_add = [
            pd.Series({**r.to_dict(), **dict(zip(affected_cols, values))})
            for values in affected_inv_vals_combinations
        ]
        to_add = [
            r for r in to_add if not _row_in_df(r, df_filtered)
        ]

        new_rows = pd.concat([new_rows, pd.DataFrame(to_add)])
```

Reply:
looks like a good compression, did you test this with `test_utils` ?

Reply:
Not yet, I'll test it and make a commit.

Reply:
commit pushed, please verify

Reply:
Looks good!